packages/frontend/src/features/metricsCatalog/components/visualization/MetricsVisualization.tsx

```diff
@@ -0,0 +1,241 @@
+    timeHour,
+    timeMinute,
+    timeMonth,
+    timeSecond,
+    timeWeek,
+    timeYear,
+} from 'd3-time';
+import dayjs from 'dayjs';
+import { useEffect, useMemo, type FC } from 'react';
+import {
+    CartesianGrid,
+    Legend,
+    Line,
+    LineChart,
+    ReferenceArea,
+    ResponsiveContainer,
+    Tooltip as RechartsTooltip,
+    XAxis,
+    YAxis,
+} from 'recharts';
+import MantineIcon from '../../../../components/common/MantineIcon';
+import { FORMATS, type TimeSeriesData } from './types';
+import { useChartZoom } from './useChartZoom';
+
+const tickFormatter = (date: Date) => {
+    return (
+        timeSecond(date) < date
+            ? FORMATS.millisecond
+            : timeMinute(date) < date
+            ? FORMATS.second
+            : timeHour(date) < date
+            ? FORMATS.minute
+            : timeDay(date) < date
+            ? FORMATS.hour
+            : timeMonth(date) < date
+            ? timeWeek(date) < date
+                ? FORMATS.day
+                : FORMATS.week
+            : timeYear(date) < date
+            ? FORMATS.month
+            : FORMATS.year
+    )(date);
+};
+
+type Props = {
+    metric: MetricWithAssociatedTimeDimension;
+    data: MetricsExplorerQueryResults;
+};
+
+const MetricsVisualization: FC<Props> = ({ metric, data }) => {
+    const { colors, radius, shadows, fontSizes } = useMantineTheme();
+
+    const defaultTimeDimension = metric.defaultTimeDimension;
+
+    const timeDimensionFieldId = Object.entries(data.fields).find(
+        ([_, field]) => 'timeInterval' in field,
+    )?.[0];
+
+    const metricFieldId = Object.entries(data.fields).find(([_, field]) =>
+        isMetric(field),
+    )?.[0];
+
+    const timeSeriesData: TimeSeriesData[] | null = useMemo(() => {
+        // TODO: Currently we hide the visualization if the default time dimension is not present. Address this differently.
+        if (!defaultTimeDimension) return null;
+
+        if (!timeDimensionFieldId || !metricFieldId) return null;
+        // TODO: zipping with index is not a smart idea.
+        return data.rows.map((row, index) => ({
+            date: new Date(String(row[timeDimensionFieldId].value.raw)),
+            metric: row[metricFieldId].value.raw,
+            ...(data.comparisonRows
+                ? {
+                      compareMetric:
+                          data.comparisonRows[index]?.[metricFieldId]?.value
+                              .raw,
+                  }
+                : {}),
+        }));
+    }, [
+        defaultTimeDimension,
+        timeDimensionFieldId,
+        metricFieldId,
+        data.rows,
+        data.comparisonRows,
+    ]);
+
+    const {
+        zoomState,
+        handlers: {
+            handleMouseDown,
+            handleMouseMove,
+            handleMouseUp,
+            resetZoom,
+        },
+        activeData,
+    } = useChartZoom({
+        data: timeSeriesData || [],
+    });
+
+    useEffect(() => {
+        resetZoom();
```

Comment:
now, every time the data changes, the zoom will reset @IrakliJani . Thank you for pointing it out. I feel like this is ok bc `data` is a stable reference, if I understand this correctly

Reply:
makes sense!